unified-codes/apps/data-service/src/app/User.ts

```diff
@@ -0,0 +1,34 @@
+import { JWTToken } from './JWT';
+
+export class UserError extends Error {
+  constructor(message: string) {
+    super(message);
+  }
+}
+
+export class User {
+  token: JWTToken;
+  roles: string[];
```

Comment:
Would perhaps be better that the user just stores token, and we pass the user and operation to `AuthorisationService.authorise(user, operation)`, which then uses an `AccessController` or something to match the task to the roles required, e.g. 

```
  async authorise(user: User, operation: Operation) {
    const isValidUser = this.provider.verifyIdentityToken(user.token);
    if (!isValidUser) return false;
    const isValidOperation = this.controller(user, operation);
    return isValidOperation;
  }
```

Something like that? Think it would be an added bonus that it would be quite easy to test the `AccessController` when its decoupled from the verification of the token.

Reply:
agree - the decoupling is important here. nice to have the provider verify the token and then access control is separate.